### ISSUE TITLE
[Feature Request] Manual VRAM flushing or automatic flushing after OOM error

### ISSUE THREAD

**Issue by joeyballentine** [microsoft/onnxruntime#11740]
I implement automatic CUDA out-of-memory detection in my program that detects when an image is too large and tiles it accordingly. This is done via catching when an OOM error happens and increasing the amount it splits until it is able to run properly. However, this requires being able to clear the VRAM being used before trying again, which ONNX seems to a) not do automatically, and b) not have any option of doing. I would prefer to not have to continually redefine an inference session every time I want to clear memory.

Therefore, I would like to have a method that allows the VRAM being used to process the input to be flushed.

**System information**
- ONNX Runtime version: 1.11.1

**Describe the solution you'd like**

I would like to have a method available that allows the current in-use VRAM to be cleared without having to define a new session. Or at the very least, VRAM should be flushed when an out-of-memory error happens (similar to how PyTorch handles it)

**Describe alternatives you've considered**
I attempted to redefine the session variable, yet I get a "MemoryError: bad allocation" when I attempt to do that. Besides, if I were to define a new session every time I wanted to try a new split factor, it would take a long time to initialize every time, which I would like to avoid.

**Additional context**
Basically just go look at what torch does. It clears out the VRAM automatically when it encounters a CUDA OOM, which allows the feature I described to work. NCNN does not clear automatically, however since it just uses manually defined Vulkan blob allocators, I am able to flush it myself. However, neither option is available with onnxruntime

**Comment by ytaous** [comment]
When does OOM kick in? During first few steps/exporting onnx, or runtime after few hundred steps?
Flushing memory blindly would actually impact performance.

**Comment by joeyballentine** [comment]
> When does OOM kick in? During first few steps/exporting onnx, or runtime after few hundred steps? Flushing memory blindly would actually impact performance.

The latter, which would make it very slow yes, but I'd still prefer this over not having any control

**Comment by ytaous** [comment]
You can try "torch.cuda.empty_cache()" in your code if it's CUDA OOM.
You can even add code to check current memory % before kicking it off.

**Comment by joeyballentine** [comment]
> You can try "torch.cuda.empty_cache()" in your code if it's CUDA OOM. You can even add code to check current memory % before kicking it off.

Using torch kinda defeats the point of using onnx. Also, the point of this is to be able to run bigger models & images than you have the vram for, so checking how much vram is in use before running isn't useful at all.

What would be useful is being able to know how much memory onnx will eventually use, if there's any way to get that information
